src/box/lua/config/applier/credentials.lua

```diff
@@ -613,6 +619,50 @@ local function sync_privileges(credentials, obj_to_sync)
                                     to_revoke.obj_name)
             end
         end
+
+        -- Collect updated information about privileges in the
+        -- database and recalculate the difference from the target
+        -- configuration.
+        box_privileges = box.schema[role_or_user].info(name)
+        box_privileges = privileges_from_box(box_privileges)
+        local missed_grants = privileges_subtract(config_privileges,
+                                                  box_privileges)
+
+        -- The most frequent scenario is when a privilege
+        -- couldn't be granted, because the object (space/
+        -- function/sequence) doesn't exist at the moment.
+        --
+        -- That's normal and the alert will be dropped on next
+        -- sync_privileges() call. If the object appears and the
+        -- privileges from the configuration are given, the alert
+        -- will not be issued again in this call.
+        --
+        -- If a serious error occurs on the privilege granting
+        -- (not a 'no such an object' one), then an alert of the
+        -- 'error' type is issued in the privileges_action_f().
+        --
+        -- The 'error' alerts aren't dropped here. The only way
+        -- to get rid of them is to try to re-apply the
+        -- configuration using config:reload() (or using automatic
+        -- reload from a remote config storage in Tarantool
+        -- Enterprise Edition).
+        --
+        -- Important: this code deliberately ignores the
+        -- `obj_to_sync` filter. It should be in-sync with the
+        -- :drop_if() call above.
+        for _, grant in ipairs(missed_grants) do
+            local alert = {
+                type = 'warn',
+                _trait = 'missed_privilege',
```

Comment:
Do we really need this `_trait`? Can't we just use lua regex to the alert message? The better option would be to use alert keys, if they start with some value, we can delete that in `drop_if`.

It just seems to me, that all droppable alerts should have keys.

Reply:
> Can't we just use lua regex to the alert message?

Message are for humans. It is acceptable to use them programmatically in a test, but not in an application logic. Neither inside tarantool, nor in a user code.

So, we need some machine-readable information. We need to drop alerts according to some properties and either encode them into keys somehow (and parse then) or add a field. The latter fits better for a machine processing (no need to encode and parse).

> It just seems to me, that all droppable alerts should have keys.

The keys serve the point access pattern, but they don't fit good to a bulk access. I wouldn't try to follow the 'droppable == have caller defined keys' rule.

Well, ideally we have an index based on the needed properties and find necessary alerts faster. However, I don't expect such amount of alerts that costs this complexity.